pep-0657.rst

```diff
@@ -0,0 +1,398 @@
+
+Motivation
+==========
+
+The primary motivation for this PEP is to improve the feedback presented about the location of errors to aid with debugging.
+
+Python currently keeps a mapping of bytecode to line numbers from compilation.
+The interpreter uses this mapping to point to the source line associated with
+an error. While this line-level granularity for instructions is useful, a
+single line of Python code can compile into dozens of bytecode operations
+making it hard to track which part of the line caused the error.
+
+Consider the following line of Python code::
+
+    x['a']['b']['c']['d'] = 1
+
+If any of the values in the dictionaries are ``None``, the error shown is::
+
+    Traceback (most recent call last):
+      File "test.py", line 2, in <module>
+        x['a']['b']['c']['d'] = 1
+    TypeError: 'NoneType' object is not subscriptable
+
+From the traceback, it is impossible to determine which one of the dictionaries
+had the ``None`` element that caused the error. Users often have to attach a
+debugger or split up their expression to track down the problem.
+
+However, if the interpreter had a mapping of bytecode to column offsets as well
+as line numbers, it could helpfully display::
+
+    Traceback (most recent call last):
+      File "test.py", line 2, in <module>
+        x['a']['b']['c']['d'] = 1
+        ^^^^^^^^^^^^^^^^
+    TypeError: 'NoneType' object is not subscriptable
+
+indicating to the user that the object ``x[‘a’][‘b’]`` must have been ``None``.
+This highlighting will occur for every frame in the traceback. For instance, if
+a similar error is part of a complex function call chain, the traceback would
+display the code associated to the current instruction in every frame::
+
+    Traceback (most recent call last):
+      File "test.py", line 14, in <module>
+        lel3(x)
+        ^^^^^^^
+      File "test.py", line 12, in lel3
+        return lel2(x) / 23
+               ^^^^^^^
+      File "test.py", line 9, in lel2
+        return 25 + lel(x) + lel(x)
+                    ^^^^^^
+      File "test.py", line 6, in lel
+        return 1 + foo(a,b,c=x['z']['x']['y']['z']['y'], d=e)
+                             ^^^^^^^^^^^^^^^^^^^^^
+    TypeError: 'NoneType' object is not subscriptable
+
+This problem presents itself quite a lot when passing down multiple objects to
+function calls while accessing the same attribute in them. For instance, this
+error::
+
+    Traceback (most recent call last):
+      File "test.py", line 19, in <module>
+        foo(a.name, b.name, c.name)
+    AttributeError: 'NoneType' object has no attribute 'name'
+
+With the improvements in this PEP this would show::
+
+    Traceback (most recent call last):
+      File "test.py", line 17, in <module>
+        foo(a.name, b.name, c.name)
+                    ^^^^^^
+    AttributeError: 'NoneType' object has no attribute 'name'
+
+
+These sorts of ambiguous error messages are commonplace when dealing with lines
```

Comment:
I wonder if we can add some extra spacing or some differentiation between each of these original-new traceback pairs, any ideas?

Reply:
I am not sure I understand what you would like to change, unfortunately. Maybe you can make a suggestion with the GitHub suggestion mechanism?

Reply:
Oh sorry, let me explain better. When rendered these tracebacks end up looking like this:
![image](https://user-images.githubusercontent.com/773529/117546427-5d7aa100-aff8-11eb-90bc-22983def44f8.png)
which makes it hard to tell that they're before and after.

Let me see if there's some restructured text magic we can use to better differentiate the pairs of before-pep and after-pep tracebacks.

Reply:
Taking inspiration from the pattern matching PEP maybe we can put the examples in bullet points like this:

https://github.com/pablogsal/peps/compare/657...ammaraskar:bullet-point

This makes it look like 
![image](https://user-images.githubusercontent.com/773529/117546772-1d1c2280-affa-11eb-9e53-3d4c85b6cc19.png)

which makes it easier to read in my opinion.

Reply:
Gotcha! I will add this bullets!